### DISCUSSION TITLE
how to blit a bitmap of an image with the st7789_4bit driver for the pico

### DISCUSSION THREAD

**Discussion by TLmv** [micropython/micropython#12145] [RP2040 / Pico]
I use micropython 1.91 combined with nanogui and am rewriting a script that takes a long time to set up.
and then i thought, wouldn't it be nice if i had something to show while it sets up?

so i created this RGB565 2 byte bitmap from a logo with a few primary colors but am at a loss how to succesfull use the st7789_4bit driver.

and i thought i found a working example, but am unable to adapt it to my needs.
although nanogui is not really used, i do use the color_setup file to get some settings and a reference to the ST7789_4bit driver.
i tried introducing a new framebuffer in the script itself, but ended up with memory issues or lost connection.
i'm willing to make a much smaller one if 240x320 is pushing it.

can you help me solve this?

this code was something i found but meant for another display driver [https://www.penguintutor.com/programming/picodisplayanimations](url)

```
from machine import Pin, I2C, PWM, SPI
from color_setup import ssd  # Create a display instance
from gui.core.writer import Writer, CWriter
from gui.core.nanogui import refresh  # Color LUT is updated now.
#import framebuf

spi = SPI(1,5_500_000,sck=Pin(6),mosi=Pin(11),miso=Pin(12)) 
refresh(ssd, True)

height = 360
width=240

#display_buffer = bytearray(width * height * 2)
display_buffer = bytearray(height * -(-width // 2))
def setup():
    blit_image_file ("logo.raw")

    # display.update()
    refresh(ssd)



# This is based on a binary image file (RGB565)
# with the same dimensions as the screen
# updates the global display_buffer directly

def blit_image_file (filename):

    global display_buffer

    with open (filename, "rb") as file:

        position = 0

        #while position < (width * height * 2):
        while position < (height * -(-width // 2)):

            current_byte = file.read(1)

            # if eof

            if len(current_byte) == 0:

                break

            # copy to buffer

            display_buffer[position] = ord(current_byte)

            position += 1

    file.close()



setup()

# Do nothing - but continue to display the image

while True:

    pass

```

**Comment by peterhinch** [comment]
The `ssd` object is an instance of the [ST7789 class](https://github.com/peterhinch/micropython-nano-gui/blob/aef423a53689cf70e552a1a2dd782c758e11324e/drivers/st7789/st7789_4bit.py#L48) which is derived from `framebuf.FrameBuffer`. If you look at the [framebuf docs](http://docs.micropython.org/en/latest/library/framebuf.html) you will see that it's possible to blit from one frame buffer to another. Create a buffer and a `FrameBuffer` instance, populate it with your graphic, and blit. You will need to call refresh to transfer the data to the display.
[EDIT]
The FrameBuffer `blit` method is used internally by `writer.py` to render glyphs, see [code](https://github.com/peterhinch/micropython-nano-gui/blob/aef423a53689cf70e552a1a2dd782c758e11324e/gui/core/writer.py#L295).

**Comment by TLmv** [comment]
thank you very much for your input, reading your post it sounds very easy and straight forward, but for me it is as hard as struggling a bear.

i did read the docs and tried a lot with framebuffers and smaller size images since i'm not sure the bitmap is correct...i have no method or tool to check that.
however since the pico and/or thonny keeps locking up i've no way to know why this happens. it appears to happen during blitting, but can't even be sure about that.
since it was only nice and not compulsory for my program, i think it's time to move on from this idea and again focus on the real purpose of the program instead of some visual flair.
